In [1]:
from __future__ import print_function
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
# from torch.utils.data import DataLoader, random_split

In [31]:
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        
        self.conv1 = nn.Sequential(
            nn.Conv2d(1, 16, 3, padding=1),
            nn.ReLU(),
            nn.BatchNorm2d(16),
            nn.Conv2d(16, 16, 3, padding=1),
            nn.ReLU(),
            nn.BatchNorm2d(16),
            nn.Conv2d(16, 16, 3, stride=2, padding=1),
            nn.ReLU(),
            nn.BatchNorm2d(16),
            nn.MaxPool2d(2, 2),
            nn.Dropout(0.25)
        )
        
        self.conv2 = nn.Sequential(
            nn.Conv2d(16, 16, 3, padding=1),
            nn.ReLU(),
            nn.BatchNorm2d(16),
            nn.Conv2d(16, 32, 3, padding=1),
            nn.ReLU(),
            nn.BatchNorm2d(32),
            nn.Conv2d(32, 32, 3, stride=2, padding=1),
            nn.ReLU(),
            nn.BatchNorm2d(32),
            nn.MaxPool2d(2, 2),
            nn.Dropout(0.25)
        )
        
        self.fc = nn.Sequential(
            nn.Linear(64, 10)
        )
        
    def forward(self, x):
        x = self.conv1(x)
        x = self.conv2(x)
        # x = self.conv3(x)
        
        x = x.view(x.size(0), -1)
        # x = self.fc(x)
        # x = nn.GovalAvgPooling(x)
        x = F.log_softmax(x, dim=1)
        return x

In [32]:
!pip install torchsummary
from torchsummary import summary
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
model = Net().to(device)
summary(model, input_size=(1, 28, 28))

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 16, 28, 28]             160
              ReLU-2           [-1, 16, 28, 28]               0
       BatchNorm2d-3           [-1, 16, 28, 28]              32
            Conv2d-4           [-1, 16, 28, 28]           2,320
              ReLU-5           [-1, 16, 28, 28]               0
       BatchNorm2d-6           [-1, 16, 28, 28]              32
            Conv2d-7           [-1, 16, 14, 14]           2,320
              ReLU-8           [-1, 16, 14, 14]               0
       BatchNorm2d-9           [-1, 16, 14, 14]              32
        MaxPool2d-10             [-1, 16, 7, 7]               0
          Dropout-11             [-1, 16, 7, 7]               0
           Conv2d-12             [-1, 16, 7, 7]           2,320
    

In [33]:


torch.manual_seed(1)
batch_size = 128

kwargs = {'num_workers': 1, 'pin_memory': True} if use_cuda else {}
train_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=True, download=True,
                    transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])),
    batch_size=batch_size, shuffle=True, **kwargs)
test_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=False, transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])),
    batch_size=batch_size, shuffle=True, **kwargs)


In [34]:
print("Train datset size =",len(train_loader))
print("Test datset size =",len(test_loader))

Train datset size = 469
Test datset size = 79


In [35]:
from tqdm import tqdm
def train(model, device, train_loader, optimizer, epoch):
    model.train()
    pbar = tqdm(train_loader)
    for batch_idx, (data, target) in enumerate(pbar):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        pbar.set_description(desc= f'loss={loss.item()} batch_id={batch_idx}')


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.4f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))

In [36]:

model = Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)

for epoch in range(1, 20):
    print("################# Epoch =",str(epoch),"#######################")
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)

################# Epoch = 1 #######################


loss=1.560489535331726 batch_id=468: 100%|██████████| 469/469 [00:23<00:00, 19.79it/s]



Test set: Average loss: 0.2668, Accuracy: 9652/10000 (96.5200%)

################# Epoch = 2 #######################


loss=1.3287440538406372 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 23.86it/s]



Test set: Average loss: 0.1891, Accuracy: 9730/10000 (97.3000%)

################# Epoch = 3 #######################


loss=1.2570136785507202 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 23.85it/s]



Test set: Average loss: 0.1237, Accuracy: 9779/10000 (97.7900%)

################# Epoch = 4 #######################


loss=1.2693179845809937 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 25.12it/s]



Test set: Average loss: 0.1204, Accuracy: 9785/10000 (97.8500%)

################# Epoch = 5 #######################


loss=1.2474802732467651 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 24.58it/s]



Test set: Average loss: 0.0990, Accuracy: 9777/10000 (97.7700%)

################# Epoch = 6 #######################


loss=0.8580544590950012 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 24.00it/s]



Test set: Average loss: 0.0988, Accuracy: 9805/10000 (98.0500%)

################# Epoch = 7 #######################


loss=0.9114904999732971 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 23.85it/s]



Test set: Average loss: 0.0802, Accuracy: 9835/10000 (98.3500%)

################# Epoch = 8 #######################


loss=1.008976936340332 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 24.79it/s]



Test set: Average loss: 0.0786, Accuracy: 9837/10000 (98.3700%)

################# Epoch = 9 #######################


loss=1.1285676956176758 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 24.93it/s]



Test set: Average loss: 0.0731, Accuracy: 9839/10000 (98.3900%)

################# Epoch = 10 #######################


loss=0.817021369934082 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 23.81it/s]



Test set: Average loss: 0.0669, Accuracy: 9845/10000 (98.4500%)

################# Epoch = 11 #######################


loss=1.2466987371444702 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 23.50it/s]



Test set: Average loss: 0.0662, Accuracy: 9847/10000 (98.4700%)

################# Epoch = 12 #######################


loss=0.7432202696800232 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 24.93it/s]



Test set: Average loss: 0.0634, Accuracy: 9839/10000 (98.3900%)

################# Epoch = 13 #######################


loss=1.0250381231307983 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 24.37it/s]



Test set: Average loss: 0.0562, Accuracy: 9868/10000 (98.6800%)

################# Epoch = 14 #######################


loss=0.7294676899909973 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 24.12it/s]



Test set: Average loss: 0.0555, Accuracy: 9856/10000 (98.5600%)

################# Epoch = 15 #######################


loss=0.8795804381370544 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 24.01it/s]



Test set: Average loss: 0.0557, Accuracy: 9865/10000 (98.6500%)

################# Epoch = 16 #######################


loss=0.8284929394721985 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 24.19it/s]



Test set: Average loss: 0.0612, Accuracy: 9858/10000 (98.5800%)

################# Epoch = 17 #######################


loss=1.0923060178756714 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 24.91it/s]



Test set: Average loss: 0.0524, Accuracy: 9880/10000 (98.8000%)

################# Epoch = 18 #######################


loss=1.293684482574463 batch_id=468: 100%|██████████| 469/469 [00:20<00:00, 23.40it/s]



Test set: Average loss: 0.0531, Accuracy: 9876/10000 (98.7600%)

################# Epoch = 19 #######################


loss=1.0414435863494873 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 23.52it/s]



Test set: Average loss: 0.0499, Accuracy: 9880/10000 (98.8000%)



## Final Result after Training
Total params: 21,264
Total epochs: 20
Test set: Average loss: 0.0499, Accuracy: 9880/10000 (98.8000%)